src/Idris/Parser.idr

```diff
@@ -776,23 +776,50 @@ mutual
-  interpStr q fname idents = do
-      strBegin
-      commit
-      xs <- many $ bounds $ interpBlock <||> strLit
-      strEnd
-      pure $ toPString <$> xs
+  singlelineStr : ParseOpts -> FileName -> IndentInfo -> Rule PTerm
+  singlelineStr q fname idents
+      = do b <- bounds $ do strBegin
+                            commit
+                            xs <- many $ bounds $ (interpBlock q fname idents) <||> strLitLines
+                            strEnd
+                            the (SourceEmptyRule _) $
+                                case traverse toPStr xs of
+                                     Left err => fatalError err
+                                     Right pstrs => pure $ pstrs
+           pure $ PString (boundToFC fname b) b.val
     where
-      interpBlock : Rule PTerm
-      interpBlock = interpBegin *> (mustWork $ expr q fname idents) <* interpEnd
+      toPStr : (WithBounds $ Either PTerm (List1 String)) -> Either String PStr
+      toPStr x = case x.val of
+                      Right (str:::[]) => Right $ StrLiteral (boundToFC fname x) str
+                      Right (_:::strs) => Left "Multi-line string is expected to begin with \"\"\""
+                      Left tm => Right $ StrInterp (boundToFC fname x) tm
 
-      toPString : (WithBounds $ Either PTerm String) -> PStr
-      toPString x
+  export
+  multilineStr : ParseOpts -> FileName -> IndentInfo -> Rule PTerm
+  multilineStr q fname idents
+      = do b <- bounds $ do multilineBegin
+                            commit
+                            xs <- many $ bounds $ (interpBlock q fname idents) <||> strLitLines
+                            endloc <- location
+                            strEnd
+                            pure (endloc, toLines xs [] [])
+           pure $ let ((_, col), xs) = b.val in
+                      PMultiline (boundToFC fname b) (fromInteger $ cast col) xs
+    where
+      toLines : List (WithBounds $ Either PTerm (List1 String)) -> List PStr -> List (List PStr) -> List (List PStr)
+      toLines [] line acc = reverse (reverse line::acc)
+      toLines (x::xs) line acc
           = case x.val of
-                 Right s => StrLiteral (boundToFC fname x) s
-                 Left tm => StrInterp (boundToFC fname x) tm
+                 Left tm => toLines xs ((StrInterp (boundToFC fname x) tm)::line) acc
+                 Right (str:::[]) => toLines xs ((StrLiteral (boundToFC fname x) str)::line) acc
+                 -- FIXME: calculate the precise FC so as to improve error report for invalid indentation.
+                 Right (str:::strs@(_::_)) => toLines xs [StrLiteral (boundToFC fname x) (last strs)]
+                                                         ((map (\str => [StrLiteral (boundToFC fname x) str]) (List.drop 1 $ reverse strs))
+                                                            ++ (List.reverse ((StrLiteral (boundToFC fname x) str)::line)::acc))
```

Comment:
why is this going through each line right to left and then reversing the list instead of left to right?

Reply:
before a line wrap is met, the incoming interpolation and literal can be pushed to the left, which should be efficient.

Reply:
Is there a mesurable performance gain? If there isn't I'd suggest sticking with the more familiar pattern of splitting the head and tail of the list rather than the prefix and the last element

Reply:
I changed them into `snoc`s.